docs/rfc-1-staking-contract.adoc

```diff
@@ -0,0 +1,381 @@
+required to finish all the pending work. Authorized applications are eligible
+to slash the operator for the entire undelegation period.
+
+==== Recovering a stake
+
+Once the undelegation period finishes, anyone - even a third party - can recover
+the stake. Staking contract sets the stake as recovered and returns all
+delegated tokens to the owner. Applications shouldn't be able to slash after
+undelegation is complete, even before the stake is returned. Information about
+owner, operator, and beneficiary, is retained in the contract in case some
+application rewards are still available for withdrawal.
+
+==== Keeping information in sync
+
+To avoid expensive calls to legacy staking contracts, it is assumed that
+information in T staking contract about the staked amount is always up-to-date.
+By design, this is always the case for native T stakers but might not be the
+case for legacy KEEP/NU stakers.
+
+T staking contract should expose a function allowing to seize some amount of T
+from the operator whose stake in T contract is lower than their stake in the
+legacy staking contract. 5% of the amount seized is given to the person who
+notified about the discrepancy and the rest is burned. The amount is a
+governable parameter and can be updated at any time by the governance, with no
+governance delay.
+
+For legacy stakers, staked amount can become out-of-sync in three cases:
+
+* stake undelegated on the legacy contract,
+* stake slashed on the legacy contract,
+* stake topped-up on the legacy contract.
+
+It is expected that stake undelegation will be first performed on T staking
+contract and then on the legacy staking contract, in the same transaction.
+
+It is expected that a top-up will be first performed on the legacy staking
+contract, and then propagated to the new staking contract, in the same
+transaction. Even if it does not happen in the same transaction, this kind of
+discrepancy is not slashable given that the stake amount on the legacy contract
+is higher than the stake amount on T staking contract.
+
+In case the stake has been slashed on the legacy contract, the operator is
+required to update their information on T staking contract as soon as possible.
+In practice, with the random beacon disabled, and tBTC v1 slashing the stake
+only in case of a proven fraud that had to be committed by all operators of
+ECDSA keep, this approach is acceptable.
+
+An integral part of the staking contract should be a bot monitoring stakes and
+notifying about discrepancies. This is especially important given that the bot
+may need to voluntarily inform about discrepancies for operators that have been
+slashed to zero.
+
+==== Slashing a stake
+
+Authorized applications can slash or seize a stake. Slash operation decreases
+the stake of an operator and burns slashed tokens. Seize decreases the stake,
+burns 95% of the stake, and awards 5% to the notifier of misbehavior. 
```

Comment:
I'm a bit on the fence about keeping the 5% limit. It's still a good rule to not let any specific recipient get more than 5% of a penalty, but it might be best to enforce this on the application side instead.

If the application can seize the entire penalty amount it can reduce the impact of e.g. TBTC misbehaviour on coverage pools, as seized T could be added to the collateral.

The downside of making the application responsible for the 5% rule is that a malfunctioning application could be used to steal stakes from operators, not just mostly burn them. Being able to limit exposure by authorising less than one's full stake would be able to mitigate this somewhat, but the risk is still there.

Reply:
We also need to consider what to do when slashing legacy/mixed operators. My recommendation would be to slash T first (to avoid legacy contract calls where possible), and convert all seized KEEP/NU to T (so that applications don't need to worry about different tokens).

Reply:
> I'm a bit on the fence about keeping the 5% limit. It's still a good rule to not let any specific recipient get more than 5% of a penalty, but it might be best to enforce this on the application side instead.
> 
> If the application can seize the entire penalty amount it can reduce the impact of e.g. TBTC misbehaviour on coverage pools, as seized T could be added to the collateral.
> 
> The downside of making the application responsible for the 5% rule is that a malfunctioning application could be used to steal stakes from operators, not just mostly burn them. Being able to limit exposure by authorising less than one's full stake would be able to mitigate this somewhat, but the risk is still there.

Same here. My default was to let the application decide and, as you say, instead of burning, donate the coverage pool without minting COV tokens, for example.

The complication though is not only the security model where we block malfunctioning applications from stealing stakes but the construction of Keep staking contract. The option when the application seizes 100% of the amount without burning [is not supported there](https://github.com/keep-network/keep-core/blob/3a4f9ca6ccf4e2713cea52f401a87be49561c592/solidity/contracts/TokenStaking.sol#L413-L425). For this RFC I assumed we want to have an MVP version, good enough to launch and support random beacon, tBTC v2, and PRE, and that we'll introduce additional complexity by upgrades, if necessary.

Reply:
> We also need to consider what to do when slashing legacy/mixed operators. My recommendation would be to slash T first (to avoid legacy contract calls where possible), and convert all seized KEEP/NU to T (so that applications don't need to worry about different tokens).

Let's discuss it [here](https://github.com/threshold-network/solidity-contracts/pull/15#discussion_r703419028). I am fine with both options. Slashing T first helps avoiding legacy contract calls but slashing KEEP/NU first might easen the transition to T over time.

Reply:
> For this RFC I assumed we want to have an MVP version, good enough to launch and support random beacon, tBTC v2, and PRE, and that we'll introduce additional complexity by upgrades, if necessary.

I'm fine with this.

Reply:
Similarly to another comment I made, the 5% limit seems arbitrary and simply a continuation of KEEP's legacy parameters. We can't foresee if some future application may have stricter seizing requirements. I'd propose to add this as a per-application, governance-level parameter in the T staking contract that's set when the application is connected to the staking contract. If such applications exist at some point, KEEP legacy stakes would not be eligible to participate due to their 5% seize limit, using the mapping we agreed some weeks ago, although it doesn't seem this is currently the case (it's certainly not the case for PRE).

Reply:
We could make it so that T will be seized 100% and legacy stake is limited to the 5%. Thus applications that need the full seizure functionality would only use T stakes (with the other benefits, such as staying better synced) and how they distribute the seized tokens would be an internal matter.

Reply:
Yes, that's roughly what we agreed o with @pdyraga during last call. There's a mapping in the staking contract that specifies which apps are compatible with the possible "stake provider" (i.e., native T, KEEP legacy, NU legacy), so we'd use this to allow/restrict KEEP legacy stakes as needed.

Reply:
In the Limitations section, there is a note we do not guarantee legacy stakes will be supported in all future applications. 

I proposed to keep the limit of 5% for the first applications so that existing Keep stakers are in a good spot after the network launch and their stake is honored by all applications launching in T for some time; implicitly it means applications need to accept 5% is the limit. Then, once everyone has a chance to earn some T to migrate, we will update the staking contract and remove 5% limit. 

I think this approach has better optics for stakers and is simpler to implement for the first version of staking contract.

If we want to remove the limit now, I think we have two options:
- store a mapping of applications supporting legacy stake in the contract (as currently proposed in the draft implementation)
- provide enough information during `authorizeApplication` so that the application can decide (is it legacy stake, if so in what proportions given that we allow mixing legacy and T stake)

I think in terms of the complexity: accept 5% limit, then upgrade is easier than mapping and mapping is easier than a smart `authorizeApplication` call.